text/0000-lifecycle-descriptor.md

```diff
@@ -0,0 +1,66 @@
+# Meta
+[meta]: #meta
+- Name: Lifecycle Descriptor
+- Start Date: 08/05/2019
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+The CNB specification defines platform/lifecycle  and lifecycle/buildpack contracts. When we inevitable introduce a breaking change to either contract, we will need a way to indicate to tools like `pack` which version of each contract a given lifecycle implements. I propose adding a `lifecycle.toml` file to all lifecycles specifying these api versions. Eventually this file could also be used to provide additional metadata or indicate the availability of non-breaking additive features.
+
+# Motivation
+[motivation]: #motivation
+
+We should do this so `pack` and other platforms can gracefully handle breaking changes in the platform/lifecycle and lifecycle/buildpack APIs. This file also provides a natural extension point for a lifecycle to provide useful metadata to users/platforms or advertise non-breaking features.
+
+`pack` currently uses the lifecycle version to make inferences about the platform/lifecycle and lifecycle/buildpack API versions. However, this has drawbacks. We want `pack` to work with newer lifecycles when there are no breaking changes. By explicitly noting breaking changes in the lifecycle/platform contract we can make `pack` helpfully fail when it encounters an `api` version it isn't familiar with, but continue to work successfully w/ new lifecycles that don't introduce breaking changes.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Lifecycle can include a `lifecycle.toml` at it's root. Initially, the only supported fields will be the `api.platform`, `api.buildpack` and `lifecycle.version`.
+
+```toml
+[api]
+ platform = "1"
+ buildpack = "1"
+
+[lifecycle]
```

Comment:
Wonder if a top-level `lifecycle` field would be necessary. I guess we do the same for `buildpack.toml`, so it's probably fine.